ext/drbbdup/drbbdup.c

```diff
@@ -1590,10 +1589,14 @@ drbbdup_instrument_dups(void *drcontext, void *tag, instrlist_t *bb, instr_t *in
             }
         }
         drreg_restore_all(drcontext, bb, instr);
-    } else if (drbbdup_is_at_end(instr) && !is_last_special) {
-        drreg_restore_all(drcontext, bb, instr);
-    } else if (drbbdup_is_at_end(instr) || drbbdup_is_exit_jmp_emulation_marker(instr)) {
-        /* Ignore instruction: hide drbbdup's own markers and the rest of the end. */
+    } else if (drbbdup_is_at_end(instr)) {
+        /* i#5906: if the emulation start label is missing we might still need to restore
+         * registers for blocks that don't end in a branch or for rep-expanded blocks. */
+        if (!is_last_special || drbbdup_ilist_has_unending_emulation(bb))
+            drreg_restore_all(drcontext, bb, instr);
+    } else if (drbbdup_is_exit_jmp_emulation_marker(instr)) {
+        /* Ignore instruction: hide drbbdup's own markers and the rest of the end.
+         * Do not call drreg_restore_all either. */
```

Comment:
Could we add a test for this?  allasm_repstr has rep string instrs and is very small, so it makes a good test target and is already run for several dr$sim options: add a test running allasm_repstr with options that trigger the filter and confirm it fails without this fix and then we have a regression test?

Reply:
Currently it won't fail because it isn't an issue in the TRACE case (because it is the final copy), and it isn't an issue in the COUNT case (because the instrumentation is smaller)

Reply:
Not sure I'm following: you're just saying the current test doesn't fail b/c it doesn't do a filter?  But the request is to add a new test that does do a filter to hit the bug and have a regression test for this behavior.

Reply:
By filter, do you mean TRACE with -L0{I,D}_filter? Because that would not trigger the failure. Or do you mean L0_FILTER? The tests in #5717 depend on this fix, so those could serve as regression tests.

For a test to fail, it needs to touch flags in the non-default case (i.e., COUNT or L0_FILTER)

Reply:
> probably we could easily extend that to trigger this issue by adding more bbdup cases to the client?

Having at least 3 cases should case the flags to be preserved, hitting this bug?  if so that's a simple augmentation of that existing test and we then have a bbdup regression test?

Reply:
I have added a test based off of drbbdup-emul-test.